Fix for #2114 Support b3 trace header

dtacalau opened the pull request:
With this fix Linkerd will propagate b3 single header the same way it propagates x-b3- multi headers. If a request contains "b3" single header and "X-B3-" multi headers, the trace context from "b3" is preferred and "x-b3-" will be ignored.  "b3" is  handled the same way as finagle: extract the traceid from "b3" header, write back "X-B3-*" multi headers, and let the flow continue as if it has received "x-b3-". 

No config changes were needed, "b3" is not written on the way out, not sure if it's a good idea since finagle does not write "b3".

Also added support for 128bit traceId has been implemented with this fix because "b3" supports 128bit traceIds.

This has been tested in two ways:
- unit tests for "b3" single header parsing and handling, also this unit tests are good for testing the handling of "x-b3-" multi headers
- manual testing by sending a request with various "x-b3-" headers to a l5d router which has tracePropagator configured, and checked the behaviour hasn't changed after implementing "b3" support.

Signed-off-by: dst4096 <dst4096@gmail.com>

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+        ()
+      }
+
+      val _ = caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList) match {
```
You do not need to assign the result of this operation to _. You probably got the impression that this needs to happen because of the compile time failure indicating that you are ignoring a non null value. But here you are not

dtacalau replied:
Done, removed assignment.

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+      }
+
+      val _ = caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList) match {
+        case Some(a) =>
```
Since this is a side - effecting operation, you do not need to match on None really, as nothing useful is happening at that point.  Also,  matching on the size in this way is not very idiomatic. You can use cons matches I think. So it would look a bit like: 

```
      caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList).foreach {
        case  Nil => ///
        case  a :: Nil =>
        case  a :: b :: Nil =>
        case  a :: b :: c :: Nil =>
      }
```

dtacalau replied:
No need to change to cons matches, since I've dropped the toList

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+            case 3 =>
+              handleTraceAndSpanIds(headers, a(0), a(1))
+              handleSampled(headers, a(2))
+            case 4 =>
```
What happens if the size of the list is more than 4, say 5. I think in this case as this is a non exhaustive match, you will get an exception.  In such cases, just ensure that your match is exhaustive or simply add a default match case

dtacalau replied:
Done, added default match, cases where size of list is not as expected will just be ignored.

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+        ()
+      }
+
+      val _ = caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList) match {
```
Not sure there is a need for `.toList` here

dtacalau replied:
Done, not needed, array is fine.

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+
+    // support setting a 128bit traceid
+    if (id.traceIdHigh.isEmpty) {
+      val __ = headers.set(ZipkinTraceHeader, id.traceId.toString)
```
no need for assignign this to a val `___`

dtacalau replied:
Done, removed all assignments.

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+
+import io.buoyant.linkerd.protocol.ZipkinTrace
+import io.buoyant.linkerd.protocol.ZipkinTracePropagator
+
```
Unneccessary empty line

dtacalau replied:
Done, removed

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+import io.buoyant.linkerd.protocol.ZipkinTracePropagator
+
+import com.twitter.finagle.http.Request
+
```
And here

dtacalau replied:
Done, removed

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+
+    // expect to get the right sampled value which is 0
+    var sampler = ZipkinTrace.getSampler(req.headerMap)
+    sampler match {
```
Instead of doing explicit `assert(false)` whenever you have an empty option, I think you can simply combine the two with the contains combinator such as `assert(sampler.contains(0.0f))`. This will be a lot more concise. That stands true for all the other places, were you yous this sort of pattern

dtacalau replied:
Done, used assert(sampler.contains(0.0f)), much cleaner, thanks

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+        assert(tid.traceId.toString().equals("a3ce929d0e0e4736"))
+        assert(tid.spanId.toString().equals("00f067aa0ba902b7"))
+        assert(tid.parentId.toString().equals("5b4185666d50f68b"))
+        assert(tid.sampled.toString() == "Some(false)")
```
Use .contains here. This is an option that should contain a bool that is false so `assert(tid.sampled.contains(false))` should do.

dtacalau replied:
Done, used sampled.contains(false), much cleaner.

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+        assert(tid.parentId.toString().equals("5b4185666d50f68b"))
+        assert(tid.sampled.toString() == "Some(false)")
+      }
+      case None => {
```
I think generally, you do not need brackets surrounding the right sides of case statements unless you want to be really explicit and emphasize something

dtacalau replied:
Done, removed all surrounding brackets, C++ habits :)

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,305 @@
+
+    // expect to get the right sampled value which is 0
+    // this also checks "x-b3-sampled" has been added to request
+    var sampler = ZipkinTrace.getSampler(req.headerMap)
```
Can't this be a `val` ?

dtacalau replied:
Done, no need for var, used val

adriancole commented:
one thing to spot check in the implementation is if "b3" is consumed or not. I know some proxies pass headers, which could result in a stale "b3" value sent out on the other side. Probably not an issue here, but worth asking

dtacalau commented:
> Nice work @dtacalau! Left a few comments, mainly around code style, etc. What I am wondering however is whether this same logic + tests should be implemented for the [h2 version of the code](https://github.com/linkerd/linkerd/blob/dcf89d6a94b86e3412297b8e45974fdd8c24a67d/linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/ZipkinTracePropagator.scala).

Done, added support for b3 in h2 implementation, also I've updated unit test for h2.

adriancole commented:
PS as timing is interesting sometimes, a user of a gateway project just found what appears to be a propagation bug relating to my earlier comment. Basically verify that this product won't end up passing through either header choice unaffected, and it will save users from lengthy troubleshooting later https://github.com/spring-cloud/spring-cloud-sleuth/issues/1452

zaharidichev commented on linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/ZipkinTracePropagator.scala:
```diff
@@ -59,28 +59,121 @@ object ZipkinTrace {
+    }
+
+    headers.set(ZipkinParentHeader, id.parentId.toString)
+    headers.set(ZipkinSampleHeader, (if ((id.sampled exists { _ == true })) 1 else 0).toString)
```
can use contains here + there is unnecessary parens

dtacalau replied:
I have redone the whole block, see my other PRs.

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +58,121 @@ object ZipkinTrace {
+    }
+
+    headers.set(ZipkinParentHeader, id.parentId.toString)
+    headers.set(ZipkinSampleHeader, (if ((id.sampled exists { _ == true })) 1 else 0).toString)
```
same here

dtacalau replied:
I have redone the whole block, see my other PRs.

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,283 @@
+      case Some(tid) =>
+        assert(tid.traceId.toString().equals("64fe8b2a57d3eff7"))
+        assert(tid.spanId.toString().equals("05e3ac9a4f6e3b90"))
+        assert(tid.traceIdHigh.toString().equals("Some(80f198ee56343ba8)"))
```
again, there is no need for Some(), just check with `contains` or any other suitable combinator

dtacalau replied:
Done

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,283 @@
+        assert(tid.spanId.toString().equals("05e3ac9a4f6e3b90"))
+        assert(tid.traceIdHigh.toString().equals("Some(80f198ee56343ba8)"))
+      case None =>
+        assert(false)  // "traceId does not exist"
```
not a great pattern in my opinion. You can do option.exists(arbitrary predicate). If you want to have seprate asserts for the separate conditions, probably do an `assert(trace.NonEmpty)` + `trace.forEach(... your asserts here....)`

dtacalau replied:
redone tests, see my other PRs.

zaharidichev commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +58,121 @@ object ZipkinTrace {
+
+      caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-")) match {
+        case Some(a) =>
+          a.size match {
```
TIOLI: I know that I said that toList was not needed and it is not if you are simply matching on the size. But dont you think it would be a bit more idiomatic to do cons matches? Then you can have the extracted parts of the list be named a lot more explicitly instead of needing to index into the array?  And generally matching on size when we have all the bells and whistles to do it in more expressive ways is a bit of a code smell.

dtacalau replied:
redone it using list, see my other PR.

zaharidichev commented on linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/http/ZipkinTracePropagatorTest.scala:
```diff
@@ -0,0 +1,283 @@
+    req.headerMap.add("x-b3-sampled", "0")
+
+    val trace = ztp.traceId(req)
+    trace match {
```
Again, making use of the combinators would probably make sense here.

adleong commented on linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala:
```diff
@@ -61,28 +58,121 @@ object ZipkinTrace {
   val ZipkinSampleHeader = "x-b3-sampled"
   val ZipkinFlagsHeader = "x-b3-flags"
 
+  /**
```
Ideally we could share code between the http and h2 implementations.  There's a `router/base-http` project that is where code shared between `router/http` and `router/h2` lives.  We could do a similar thing here and create a `linkerd/protocol/base-http` project to hold code which can be shared by `linkerd/protocol/http` and `linkerd/protocol/h2`.

dtacalau replied:
Done, moved common code to linkerd/protocol/base-http and redone the whole implementation around it. See my other PR.

dtacalau commented on linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/ZipkinTracePropagator.scala:
```diff
@@ -59,28 +59,121 @@ object ZipkinTrace {
   def get(headers: Headers): Option[TraceId] = {
-    val trace = caseInsensitiveGet(headers, ZipkinTraceHeader).flatMap(SpanId.fromString)
+    //if "b3" present convert it to "x-b3-" multio and let the flow continue as if it has received "x-b3-"
+    convertB3SingleHeaderToMultiHeaders(headers)
```
Yup, feels the same to me, I've redone the implementation, see my other PR.

dtacalau commented on linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/ZipkinTracePropagator.scala:
```diff
@@ -59,28 +59,121 @@ object ZipkinTrace {
   }
 
+  /**
+   * This method will also work on requests containing "b3" single header, because after the header is read and
```
Fixed in my other PR.

dtacalau commented:
The code from this PR has been rewritten and I've created another PR
